library/core/src/any.rs

```diff
@@ -734,3 +736,45 @@
 pub const fn type_name_of_val<T: ?Sized>(_val: &T) -> &'static str {
     type_name::<T>()
 }
+
+/// Try to swap two `dyn Any` references.
+///
+/// If the two references refer to the same types, the two values are swapped
+/// and `true` is returned.
+/// If not, the nothing is done and `false` is returned.
+///
+/// # Example
+///
+/// ```rust
+/// use std::any::{Any, try_swap};
+///
+/// let mut a = String::from("a");
+/// let mut b = String::from("b");
+/// let mut c = vec!['c'];
+///
+/// // Swapping two value of the same type works
+/// assert!(try_swap(&mut a as &mut dyn Any, &mut b as &mut dyn Any));
+/// assert_eq!((&*a, &*b), ("b", "a"));
+///
+/// // Swapping two values with different types fails
+/// assert!(!try_swap(&mut a as &mut dyn Any, &mut c as &mut dyn Any));
+/// ```
+#[must_use = "`try_swap` returns `false` if it failed"]
+#[unstable(feature = "any_try_swap", issue = "120319")]
+pub fn try_swap(a: &mut dyn Any, b: &mut dyn Any) -> bool {
+    let swap = (a as &dyn Any).type_id() == (b as &dyn Any).type_id();
+
+    if swap {
+        let len = mem::size_of_val(a);
```

Comment:
Is it guaranteed somewhere that types with the same TypeId have the same size? While dyn casts currently require a type to be sized, the `Any` trait is also implemented on unsized types. So if unsized dyn casts are implemented then this is unsound. (The unsized dyn cast's implementation is not relevant to this PR :)

Reply:
Hmmm... I hadn't thought of that.

There are two questions here:
- Will it ever happen ? (Will we want that one day ?)
- If this happens, can it be supported ? (Returning `false` in case of mismatch is probably fine)